x-pack/plugin/rest-compatibility/src/main/java/org/elasticsearch/compat/CompatibleVersionPlugin.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.compat;
+
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.xcontent.MediaType;
+import org.elasticsearch.common.xcontent.ParsedMediaType;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.plugins.RestCompatibilityPlugin;
+import org.elasticsearch.rest.RestStatus;
+
+public class CompatibleVersionPlugin extends Plugin implements RestCompatibilityPlugin {
+
+    @Override
+    public Version getCompatibleVersion(
+        @Nullable ParsedMediaType acceptHeader,
+        @Nullable ParsedMediaType contentTypeHeader,
+        boolean hasContent
+    ) {
+        Byte aVersion = parseVersion(acceptHeader);
+        byte acceptVersion = aVersion == null ? Version.CURRENT.major : Integer.valueOf(aVersion).byteValue();
+        Byte cVersion = parseVersion(contentTypeHeader);
+        byte contentTypeVersion = cVersion == null ? Version.CURRENT.major : Integer.valueOf(cVersion).byteValue();
+
+        // accept version must be current or prior
+        if (acceptVersion > Version.CURRENT.major || acceptVersion < Version.CURRENT.major - 1) {
+            throw new ElasticsearchStatusException(
+                "Compatible version must be equal or less then the current version. Accept={}} Content-Type={}}",
+                RestStatus.BAD_REQUEST,
+                acceptHeader,
+                contentTypeHeader
+            );
+        }
+        if (hasContent) {
+
+            // content-type version must be current or prior
+            if (contentTypeVersion > Version.CURRENT.major || contentTypeVersion < Version.CURRENT.major - 1) {
+                throw new ElasticsearchStatusException(
+                    "Compatible version must be equal or less then the current version. Accept={} Content-Type={}",
+                    RestStatus.BAD_REQUEST,
+                    acceptHeader,
+                    contentTypeHeader,
+                    RestStatus.BAD_REQUEST
+                );
+            }
+            // if both accept and content-type are sent, the version must match
+            if (contentTypeVersion != acceptVersion) {
+                throw new ElasticsearchStatusException(
+                    "Content-Type and Accept version requests have to match. Accept={} Content-Type={}",
+                    RestStatus.BAD_REQUEST,
+                    acceptHeader,
+                    contentTypeHeader
+                );
+            }
+            // both headers should be versioned or none
+            if ((cVersion == null && aVersion != null) || (aVersion == null && cVersion != null)) {
```

Comment:
Can this ever evaluate true ? wouldn't the above if catch any this case ?

Reply:
the above if `            if (contentTypeVersion != acceptVersion) {` would not catch the scenario when one of the headers is set to `compatible-with=8` (current in general) and the other is not set (defaults to current)


```
   expectThrows(
            ElasticsearchStatusException.class,
            () -> requestWith(acceptHeader(CURRENT_VERSION), contentTypeHeader(null), bodyPresent())
        );
```

```
expectThrows(
            ElasticsearchStatusException.class,
            () -> requestWith(acceptHeader(null), contentTypeHeader(CURRENT_VERSION), bodyPresent())
        );
```